service/controller/v18/resource/ipam/create.go

```diff
@@ -104,13 +128,29 @@ func (r *Resource) allocateSubnet(ctx context.Context) (string, error) {
 
 		subnet, err = ipam.Free(r.networkRange, r.allocatedSubnetMask, reservedSubnets)
 		if err != nil {
-			return "", microerror.Maskf(err, "networkRange: %s, allocatedSubnetMask: %s, reservedSubnets: %#v", r.networkRange.String(), r.allocatedSubnetMask.String(), reservedSubnets)
+			return net.IPNet{}, microerror.Maskf(err, "networkRange: %s, allocatedSubnetMask: %s, reservedSubnets: %#v", r.networkRange.String(), r.allocatedSubnetMask.String(), reservedSubnets)
 		}
 
 		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found free subnet %#q", subnet.String()))
 	}
 
-	return subnet.String(), nil
+	return subnet, nil
+}
+
+func (r *Resource) selectRandomAZs(n int) ([]string, error) {
+	if n > len(r.availabilityZones) {
+		return nil, microerror.Maskf(invalidParameterError, "requested nubmer of AZs %d is bigger than number of available AZs %d", n, len(r.availabilityZones))
+	}
+
+	// availabilityZones must be copied so that original slice doesn't get shuffled.
+	shuffledAZs := make([]string, len(r.availabilityZones))
+	copy(shuffledAZs, r.availabilityZones)
```

Comment:
I would find this being less magic. One line instead of two. No explicit allocation. No magic copy. 

```
shuffledAZs := append([]string{}, r.availabilityZones...)
```

Reply:
I wouldn't say stdlib `copy` is any kind of magic as this explicitly states what is happening/needed, but I'm fine with your suggestion as well.

Reply:
I find `shuffledAZs := append([]string{}, r.availabilityZones...)` more magic. When using `copy` the intention is clear.

Reply:
I'll keep `copy` then.